### ISSUE TITLE
Optimizer does not change weights

### ISSUE THREAD

**Issue by homocomputeris** [google/jax#2097]
I want to train a simple binary classifier in JAX STAX:
```python
import jax.numpy as np
from jax import grad, jit, random
from jax.experimental import optimizers, stax
from jax.experimental.stax import Dense, Relu, Sigmoid
from sklearn.datasets import make_circles
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


def prepare_circles(n_samples):
    X, y = make_circles(n_samples, noise=0.2, factor=0.5, random_state=1)
    X = StandardScaler().fit_transform(X)
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.4, random_state=42
    )
    return X_train, X_test, y_train, y_test


learning_rate = 0.01
n_epochs = 100
n_features = 2
n_hidden_layers = 1
n_nodes = 4
n_samples = 1000

X_train, X_test, y_train, y_test = prepare_circles(n_samples)

init_fun, apply_fun = stax.serial(
    Dense(n_nodes), Relu, Dense(n_nodes), Relu, Dense(1), Sigmoid
)
out_shape, params = init_fun(random.PRNGKey(2), (n_samples, n_features))
print(params)

opt_init, opt_update, get_params = optimizers.adam(step_size=learning_rate)
opt_state = opt_init(params)


def loss(params, x, y):
    p = apply_fun(params, x)
    ce_loss = -np.sum(y * np.log(y) + (1 - y) * np.log(1 - y))
    return ce_loss


# Define a compiled update step
@jit
def step(i, opt_state, x, y):
    params = get_params(opt_state)
    return opt_update(i, grad(loss)(params, x, y), opt_state)


for i in range(n_epochs):
    opt_state = step(i, opt_state, X_train, y_train)

params = get_params(opt_state)
print(params)
```

The problem is that the weights seem to be not updated at all.
Is it a bug or am I missing something?

**Comment by hawkinsp** [comment]
Thanks for the issue report!

`grad(loss)(params, x, y)` takes the derivative of `loss` with respect to `params`, but your loss function doesn't actually depend on the parameters (only on `y`).

```
def loss(params, x, y):
    p = apply_fun(params, x)
    ce_loss = -np.sum(y * np.log(y) + (1 - y) * np.log(1 - y))
    return ce_loss
```

Did you mean to use `p` in `loss`?

Does that answer your question?

**Comment by homocomputeris** [comment]
Yep, obviously my bad. Thanks!
